Design/3093-graphql-defer.md

```diff
@@ -0,0 +1,356 @@
+
+```swift
+// Sample code in MultipartResponseParsingInterceptor
+public struct MultipartResponseParsingInterceptor: ApolloInterceptor {
+  private static let multipartParsers: [String: MultipartResponseSpecificationParser.Type] = [
+    SubscriptionResponseParser.protocolSpec: SubscriptionResponseParser.self,
+    DeferResponseParser.protocolSpec: DeferResponseParser.self
+  ]
+
+  public func interceptAsync<Operation>(
+    chain: RequestChain,
+    request: HTTPRequest<Operation>,
+    response: HTTPResponse<Operation>?,
+    completion: @escaping (Result<GraphQLResult<Operation.Data>, Error>) -> Void
+  ) where Operation : GraphQLOperation {
+    // response validators not shown
+
+    guard
+      let multipartBoundary = response.httpResponse.multipartBoundary,
+      let protocolSpec = response.httpResponse.multipartProtocolSpec,
+      let protocolParser = Self.multipartParsers[protocolSpec]
+    else {
+      // call request chain error handler
+
+      return
+    }
+
+    let dataHandler: ((Data) -> Void) = { data in
+      // proceed ahead on the request chain
+    }
+
+    let errorHandler: (() -> Void) = {
+      // call request chain error handler
+    }
+
+    protocolParser.parse(
+      data: response.rawData,
+      boundary: multipartBoundary,
+      dataHandler: dataHandler,
+      errorHandler: errorHandler
+    )
+  }
+}
+
+// Sample protocol for multipart specification parsing
+protocol MultipartResponseSpecificationParser {
+  static var protocolSpec: String { get }
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  )
+}
+
+// Sample implementations of multipart specification parsers
+
+struct SubscriptionResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "subscriptionSpec=1.0"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // parsing code currently in MultipartResponseParsingInterceptor
+  }
+}
+
+struct DeferResponseParser: MultipartResponseSpecificationParser {
+  static let protocolSpec: String = "deferSpec=20220824"
+
+  static func parse(
+    data: Data,
+    boundary: String,
+    dataHandler: ((Data) -> Void),
+    errorHandler: (() -> Void)
+  ) {
+    // new code to parse the defer specification
+  }
+}
+```
+
+#### Response data
+
+The initial response data and data received in each incremental response will need to be retained and combined so that each incremental response can insert the latest received incremental response data at the correct path and return an up-to-date response to the request callback.
+
+The data being retained and combined should not require another pass through the GraphQL executor though.
```

Comment:
Really? I would assume it would have to go back through the executor. How do you plan on merging it without execution?

Reply:
This was mentioned purely from an efficiency point of view. If the implementation of doing that is overly complicated then yes each response will have to be executed on.